### DISCUSSION TITLE
How to run GLSP as a Eclipse Plugin

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1039] [General]
Hi, 
I want to provide my current project [Open-BPMN](https://www.open-bpmn.org/) also as a Eclipse Plugin. 
The [glsp-eclipse-integration](https://github.com/eclipse-glsp/glsp-eclipse-integration) project looks to me like an example how to build a Eclipse Tool as a separate Target Platform Project. 
This seems to be oversized to me. I just want to give Eclipse Java developers a plugin (simmilar to a VS-Code extension) to add a new editor to their IDE. 

From the examples I did not find a starting point how to test the Eclipse integration as a single plugin without the need to build a complete new target platform?

Can someone give me a tip how to bundle my glsp server and client part as a single plugin?

**Comment by tortmayr** [comment]
Hi @rsoika,
sorry for the late reply.
I think you are confusing Eclipse target platform definitions with Eclipse products. The target platform is just a tool for development time that gives us an easy way to specify the required plugins etc. for our Eclipse integration plugins. It more a less just an easy way for developers to setup their Eclipse instance with one click ("set target platform") instead of having to install all required plugins manually.
This does not affect the built artifacts. The "Workflow Example Editor` for instance is still just a single plugin that can be installed in any (compatible) Eclipse installation via update site e.g:

![image](https://github.com/eclipse-glsp/glsp/assets/2311075/7913b337-dd62-4a42-b88e-0ca79132d9c8)

Nevertheless if you don't want to use a target platform you can also install the required plugins manually and then start developing your glsp eclipse editor plugin (e.g. your version of `org.eclipse.glsp.examples.workfloweditor`).
If you are a 1.1.0 Release Candiate the prerequistes are
- Eclipse 2021-09 or higher
- The "M2E- PDE Integration" plugin must be installed (GLSP uses a jetty update site that can only be consumed with this plugin)
  This plugin should be availalbe in the official release repo (e.g: https://download.eclipse.org/releases/2022-12/)
- The `GLSP Eclipse IDE Integration` plugin must be installed 
  (e.g.  https://download.eclipse.org/glsp/ide/p2/staging/1.1.0.RC10/)

**Reply by rsoika** [reply]
Thanks for your response. I think now I got it :-)

**Comment by rsoika** [comment]
Hi @tortmayr, now I started to setup my own Open-BPMN-Eclipse-Integration project. 

As far as I understand I need a client and server module. The client part seems clear to me in the moment. But the server part confuses me.

In the glsp-eclipse-integration project you have a folder `/server/example/org.eclipse.glsp.ide.workflow.editor`. This seems to me is the project specific part and I mapped this so something like `/server/open-bpmn/org.open-bpmn.bpmn.editor`. This seems to be the place where I can put my project specific server code like the `Activator.java` or the `BPMNServerManger.java` 

So far I hope I still on the right path.

But the `/plugins/` folder is not clear to me. This looks to me like this is the GLSP generic part? And I should NOT copy this folder into my project - right?

**Reply by tortmayr** [reply]
Hi @rsoika,
yes your understanding is correcth the `org.eclipse.glsp.ide.workflow.editor` plugin in the example folder is the plugin providing the custom GLSP editor. The generic integration plugins (in the "plugin" folder) should no be part of your project and will be installed via update site.
We also have a project template for the eclipse integration https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse. You should structure your project in  a similar fashion.

**Reply by rsoika** [reply]
Hi @tortmayr,
I am still struggling with the project setup. 
When I follow the `project-templates/java-emf-eclipse` then my understanding is, that the module `org.eclipse.glsp.example.javaemf.server` again is not needed in my case? I have already implemented a GLSP Server. I have all these DiagramModules, Model classes and CreationHandlers.

So I only need a layout like this:

```
.open-bpmn-eclipse-integration
├── client
│   └── open-bpmn-webapp
│       ├── app
│       ├── css
│       └── src
└── server
    ├── open-bpmn.eclipse.editor
    │   ├── diagram
    │   ├── META-INF
    │   └── src
    └── open-bpmn.eclipse.target
```



 right?

**Reply by tortmayr** [reply]
Correct. The layout looks good to me. The `org.eclipse.glsp.example.javaemf.server` plugin is the language specific integration plugin for the tasklist GLSP editor. In our case this corresponds to the `open-bpmn-eclipse-editor` plugin.

**Reply by rsoika** [reply]
Hi @tortmayr 
hopefully my last question regarding the project setup today ;-)
Do you think I need to add the [server part](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.glsp-server) of my Open-BPMN project as a maven dependency into the eclipse plugin project, or should I add the jar file into a separate /lib folder and configure it via the Eclipse Plugin Build Configuration? The later one would be more similar to the setup of the VSCode integration. In that case I would manually copy the `open-bpmn.server-1.0.2-SNAPSHOT-glsp.jar` from my server project into the /lib/ folder of the Eclipse Integration Plugin project instead of using a dependency?

I know that this is more a Eclipse specific question rather than a GLSP topic. But I am a little bit stumped at the moment.

I also think that my project is a bit unusual from your point of view. I practically build the Open-BPMN project as a Theia solution and then I have two separate projects for the VSCode and Eclipse integration. I guess most users build GLSP projects for exactly one of these target platforms? But providing one Editor for different IDEs is one of the coolest features of GLSP!

**Comment by tortmayr** [comment]
>Do you think I need to add the [server part](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.glsp-server) of my Open-BPMN project as a maven dependency into the eclipse plugin project, or should I add the jar file into a separate /lib folder and configure it via the Eclipse Plugin Build Configuration? The later one would be more similar to the setup of the VSCode integration. In that case I would manually copy the open-bpmn.server-1.0.2-SNAPSHOT-glsp.jar from my server project into the /lib/ folder of the Eclipse Integration Plugin project instead of using a dependency?

I don't think that just adding the server part as maven dependency would be enough. You still have to make the server code somehow available as an Eclipse plugin.
I personally would go for the second approach and manually copy the server into a `lib` folder. You could also try to automate this process somehow. e.g. by adding it as a maven dependency and then use the [maven-dependency-plugin](https://maven.apache.org/plugins/maven-dependency-plugin/examples/copying-artifacts.html) to copy the artifact into your lib folder.

>I also think that my project is a bit unusual from your point of view. I practically build the Open-BPMN project as a Theia solution and then I have two separate projects for the VSCode and Eclipse integration. I guess most users build GLSP projects for exactly one of these target platforms? But providing one Editor for different IDEs is one of the coolest features of GLSP!

Well, you are partly right. Currently most adopters build their project for one specif target platform/application context. Nevertheless, we also think that possibility to develop the core glsp editor platform independent and enable integration for different IDEs/platforms  with a rather slim layer of glue code is one of the strong suits of GLSP. 

IMO the theia and vscode integration already support this pattern very well by maintaining a strong separation between the generic GLSP client package and a platform specific package for customization/glue code. Here the default approach is also to start the server in a separate process which makes it easy to reuse an existing server implementation without much additional effort.

The Eclipse Integration in its current state only works with Java-based servers and operates on the assumption that the server is available as Eclipse Plugin. This enables easy integration with the Eclipse Extension API. In the long run we want to refactor the project to be more in line with other integration projects i.e. enable reuse of any existing server implementation (Java or Node) but currently we don't have any resources for that. We rely on sponsored contributions for more sophisticated changes to the Eclipse Integration other than bug fixes and regular updates to new GLSP versions.

TLDR: Sharing/Reusing GLSP editors between different platforms is a very important use case for us and we continously work on improving the process and try to reduce the overall effort needed. Please let us now if you encounter any other issues that could be improved in this context.

**Comment by rsoika** [comment]
Hi @tortmayr ,
I have made a lot of progress and finally I think I found the correct project setup to build my very basic protoype.

https://github.com/imixs/open-bpmn-eclipse-integration

I can build my project with  `yarn build`  and I can launch the Plugin with my Open-BPMN.launch configuration.

Of course now a lot of errors occur. The root of the Exceptions seems to be inside the `GLSPServerManager` or my Subclass. 

Now the problem is that I struggle to debug this because of the unclear version I should use. The project template java-emf-eclipse seems to be very different from the current master branch of glsp-eclipse-integration. It looks to me that the API of the `GLSPServerManager` has changed somehow?

My question is: what should I do? Should I checkout a specific version of the glsp-eclipse-integration to continue with debugging (which version is the correct one?) or should I change the impl of my TaskListServerManager to be compatible with the glsp-eclipse-integration master branch?

**Comment by rsoika** [comment]
Hi @tortmayr ,
now I solved most of the setup problems. I can build and start the plugin.

But when I open a .bpmn file I only get a blank swt browser window. No message on the log :-(
I have no idea how to debug now to get an understanding what is missing or is wrongly configured.

https://github.com/imixs/open-bpmn-eclipse-integration

Can you give me a tip what I should do? The funny thing - and maybe this is the only hint for the root of the problem - I can still open .wf files and the workflow editor opens up, even when I not have configured it.

**Reply by rsoika** [reply]
Maybe it is something with the `getGlspId` method in the `GLSPServerManager`? But I can't find a glimpse of a comment regarding this `GlspId`.
I know you avoid writing too many comments in the code, but maybe this ID could have some meaning ? ;-)

**Comment by rsoika** [comment]
After more analysis I came to the conclusion that the web module in the eclipse `/diagram/` folder is the root of the problem. 
I tested this by replacing the diagram.html page with a hello-world page on the Eclipse Plugin side and I could see that the browser loads the page correctly. This means to me, that the client I build was incorrect.

So from that assumption, I now re-started the client part once again from scratch and followed the `glsp-examples/project-templates/java-emf-eclipse/` once again line by line.

... but now my yarn build now fails with the error:

```
...
@open-bpmn/open-bpmn-eclipse-webapp: $ tsc && webpack
@open-bpmn/open-bpmn-eclipse-webapp: src/app.ts(50,50): error TS2339: Property 'createWebsocketConnectionProvider' does not exist on type 'typeof JsonrpcGLSPClient'.
```

Maybe this is from my devDepencency `"@eclipse-glsp/config": "1.1.0-next.d03ad9f.112"` or that my GLSP Client is based on `"@eclipse-glsp/client": "1.1.0-RC04" `?

So again I am lost in the node.js dependency hell :-(

**Reply by rsoika** [reply]
Finally I managed to resolve all dependency issues after upgrading to `1.1.0-RC10` :-)
The yarn process seems now to be correct building the `bundle.js` and `bundle.js.map` file 

With my integration code I went back to the `glsp-eclipse-integration` project which seems to be newer. 

Anyway after all I have the same result. The integrated eclipse browser opened a blank page. If I could only see the javaScript console log from this window.....

**Comment by rsoika** [comment]
Hi @tortmayr,
I invested much more hours now but I get it not working. The editor window opens but is still blank. 

From all my analyses I think the problem is the bundle.js file which I created with yarn. With some tricks I tested the resoruce URL in an external browser to get a JavaScript Console. When I compare the outcome with the GLSP Workflow example (which works fine) I can see the following javaScript errors:


```
GET
http://localhost:34547/favicon.ico
[HTTP/1.1 404 Not Found 23ms]

Uncaught TypeError: class heritage _glsp_sprotty_1.SChildElement is not an object or null
    node_modules eclipse-glsp/ide/node_modules/@eclipse-glsp/client/lib/features/change-bounds/model.js@http://localhost:34547/bundle.js line 29 > srcScript:23028
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/node_modules/@eclipse-glsp/client/lib/features/accessibility/view-key-tools/deselect-key-tool.js@http://localhost:34547/bundle.js line 29 > srcScript:21704
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/node_modules/@eclipse-glsp/client/lib/features/accessibility/view-key-tools/di.config.js@http://localhost:34547/bundle.js line 29 > srcScript:21788
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/node_modules/@eclipse-glsp/client/lib/features/accessibility/di.config.js@http://localhost:34547/bundle.js line 29 > srcScript:20638
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/node_modules/@eclipse-glsp/client/lib/index.js@http://localhost:34547/bundle.js line 29 > srcScript:31753
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/lib/features/copy-paste/di.config.js@http://localhost:34547/bundle.js line 29 > srcScript:18645
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    node_modules eclipse-glsp/ide/lib/index.js@http://localhost:34547/bundle.js line 29 > srcScript:18804
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    js http://localhost:34547/bundle.js line 29 > srcScript:157312
    __webpack_require__ http://localhost:34547/bundle.js line 29 > srcScript:158383
    <anonymous> http://localhost:34547/bundle.js line 29 > srcScript:158538
    <anonymous> http://localhost:34547/bundle.js line 29 > srcScript:158540
    <anonymous> http://localhost:34547/bundle.js line 29 > srcScript:158542
[model.ts:50](webpack://open-bpmn/node_modules/@eclipse-glsp/ide/node_modules/@eclipse-glsp/client/src/features/change-bounds/model.ts)
    js model.ts:50
    Webpack 18
```

Can you see from this output something which gives you an idea what I am doing wrong in my webpack?

Is there a chance that I can test the webpack part with an external server in VSCode ?

**Comment by tortmayr** [comment]
@rsoika Sorry for the delayed response. I somehow missed the notifications for this thread.
Great news that you managed to progress this far. 
>Can you see from this output something which gives you an idea what I am doing wrong in my webpack?

Unfortunately I have never encountered this error and from the output alone I can not make an educated guess about the root cause.
I will have a more detailed look at your implementation later today.

**Reply by rsoika** [reply]
thanks for your response. My main problem is that I have no debugging option. I think something in my web module is not correctly build :-/
The general Eclipse-Plugin Mechanism seems no longer be the problem

**Reply by tortmayr** [reply]
>Thanks for your response. My main problem is that I have no debugging option. I think something in my web module is not correctly build :-/

Could you eloborate on that? If you are able to open the diagram in an external Browser, you should also be able to debug it. Or am I missing something here?

**Reply by rsoika** [reply]
My problem is, that I do not know how I can open the webpack in an external browser. And when I open it in the SWT Browser we have no debug tool :-(

**Reply by rsoika** [reply]
I think it actually has something to do with the different releases. I will now first update from RC04 to RC10 and then I will it try again....

**Comment by rsoika** [comment]
Hi @tortmayr , I know I'm annoying, but I do not know what to do :-(

- I have upgraded my Open-BPMN Project to GLSP Version 1.1.0-RC10 
- Next, I have upgraded my Open-BPMN-VSCode integration project also to 1.1.0-RC10 

Both parts are now working fine. So with my core project I think I am on the right way.

But Now I am back to the initial problem: how to write an Eclipse Integration project? And its still the version which are confusing me. 
If I setup my project similar to the [EMF Eclipse Tempalte](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse) (as we discussed above) I still have the problem building the project.

You say, the _Eclipse target platform definitions_ are just to give me an easy way to develop my plugin. But I do not see this. 
I am currently working with the latest Eclipse version 2023-06. 

The [documentation](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse) says:

> To resolve all compilation errors you have to set the correct target platform. Got to the preferences (Window -> Preferences) and navigate to the target platform options (Plug-in Development -> Target Platform)- Selected the 2021-03-Release target definition and click Apply. Once the target platform has been resolved there should be no more compilation errors.

This does not work (or maybe my eclipse is now totally corrupted). Is the emf-example template really based on 1.1.0-RC10? The only way to get my compilation errors resolved is to import the [GLSP Eclipse Integration Project version 1.1.0-RC10](https://github.com/eclipse-glsp/glsp-eclipse-integration/tree/v1.1.0.RC10) into my workspace.

In my Editor `MANIFEST.MF` I have now referred to RC10 but this does not work and maybe it is totally wrong?

![image](https://github.com/eclipse-glsp/glsp/assets/1339124/c3967fa1-9873-4966-aa10-667824370cc3)

I still trying to find a mix between GLSP-Eclipse-Integration, GLSP-Examples, and GLSP-Server.....

**Reply by rsoika** [reply]
Hi @tortmayr ,
after more investigation I now believe, that I run into a general problem in the overall concept. 
At the first place I tried (as mentioned above) to add my existing Open-BPMN-GLSP Server jar into the folder `/lib/`. I copied my glsp fatjar version `open-bpmn.server-1.0.3-SNAPSHOT-glsp.jar`. But this jar includes all the GLSP-Server stuff from version 1.1.0.RC10. For the VSCode integration this was fine. But I believe now that for the Eclipse integration this is the reason for all my compiler and runtime errors regading class name clashes and ClassLoading exceptions.

So I replaced the jar `open-bpmn.server-1.0.3-SNAPSHOT-glsp.jar` with the jar `open-bpmn.server-1.0.3-SNAPSHOT.jar` which is only holding my custom DiagramModule implementation. 

What is your opinion about this? Do you think this should work? And than the question is: which libraries need to be added to my eclipse project? At the moment it looks to me that I could not found a Eclipse Integration version based on GLSP 1.1.0.RC10 as you can see in the screenshot above.

**Reply by tortmayr** [reply]
Hi @rsoika,
sorry for the late reply.
>So I replaced the jar open-bpmn.server-1.0.3-SNAPSHOT-glsp.jar with the jar open-bpmn.server-1.0.3-SNAPSHOT.jar which is only holding my custom DiagramModule implementation.

Yes I think this is definitively the way to go for this case. 

>And than the question is: which libraries need to be added to my eclipse project? At the moment it looks to me that I could not found a Eclipse Integration version based on GLSP 1.1.0.RC10 as you can see in the screenshot above.

The Eclipse Integration version is derived from the update sites you provided in your target file definition. If you want to use R10 you have to update the target file accordingly to `https://download.eclipse.org/glsp/ide/p2/staging/1.1.0.RC10/`.

**Comment by rsoika** [comment]
Finally it worked!!! :+1: :rocket: :birthday: :octopus: 

After my third(!) attempt, I finally made it. My Eclipse integration is now based on GLSP v2.1.0.
I refactored a lot of things again. And I think the main problem back then was the intermediate versions of GLSP 1.1.x that I had used before. 

Now it looks relatively promising and to be honest I didn't expect to ever get it done. 

However, the discussion we have had here is still important and could help others with the integration into Eclipse.
